formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonDecoder.kt

```diff
@@ -19,13 +19,23 @@ import kotlin.jvm.*
 internal open class StreamingJsonDecoder(
     final override val json: Json,
     private val mode: WriteMode,
-    @JvmField internal val lexer: JsonLexer
+    @JvmField internal val lexer: JsonLexer,
+    private val descriptor: SerialDescriptor
 ) : JsonDecoder, AbstractDecoder() {
 
     override val serializersModule: SerializersModule = json.serializersModule
     private var currentIndex = -1
     private val configuration = json.configuration
 
+    // TODO move to data class? then need to instantiate it of configuration.omitNull == true
+    private var lowerReadMark: Long = 0
+    private var highReadMarks: LongArray? = null
+    private var dummyNull: Boolean = false
```

Comment:
It looks like that all these values and methods depend only on `SerialDescriptor`, so it should be easy to move them to a separate stateful class and instantiate it here. Probably we could even reuse this in protobuf, although we don't have to.

Reply:
Doesn't the additional memory allocation decrease the performance for `omitNull = true`?

Reply:
Additional memory allocation for one object? I don't think so. You're still allocating a `highReadMarks` every time here.

In future, when we implement `SerialDescriptorLocal`, we can store and reuse arrays there

Reply:
`highReadMarks` created only if an object has more than 64 fields - it's a rare case